rust/kernel/sync/arc.rs

```diff
@@ -7,178 +7,216 @@
-            if obj.get_count().count.fetch_sub(1, Ordering::Release) != 1 {
+            let inner = unsafe { self.ptr.as_ref() };
+
+            // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
+            // this instance is being dropped, so the broken invariant is not observable.
+            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount. We
+            // also need to limit the scope of `obj` as it cannot be accessed after the refcount is
+            // decremented.
+            let is_zero = unsafe { rust_helper_refcount_dec_and_test(inner.refcount.get()) };
+            if !is_zero {
                 return;
             }
         }
 
-        // Synchronises with the release when decrementing above. This ensures that modifications
-        // from all previous threads/CPUs are visible to the underlying object's `drop`.
-        fence(Ordering::Acquire);
-
         // The count reached zero, we must free the memory.
         //
-        // SAFETY: The pointer was initialised from the result of `Box::into_raw`.
+        // SAFETY: The pointer was initialised from the result of `Box::leak`.
         unsafe { Box::from_raw(self.ptr.as_ptr()) };
     }
 }
 
-/// Trait for reference counted objects.
+/// A borrowed [`Ref`] with manually-managed lifetime.
 ///
-/// # Safety
-///
-/// Implementers of [`RefCounted`] must ensure that all of their constructors call
-/// [`Ref::try_new`].
-pub unsafe trait RefCounted {
-    /// Returns a pointer to the object field holds the reference count.
-    fn get_count(&self) -> &RefCount;
-}
-
-/// Holds the reference count of an object.
+/// # Invariants
 ///
-/// It is meant to be embedded in objects to be reference-counted, with [`RefCounted::get_count`]
-/// returning a reference to it.
-pub struct RefCount {
-    count: AtomicUsize,
+/// There are no mutable references to the underlying [`Ref`], and it remains valid for the lifetime
+/// of the [`RefBorrow`] instance.
+pub struct RefBorrow<T: ?Sized> {
+    inner_ref: ManuallyDrop<Ref<T>>,
 }
 
-impl RefCount {
-    /// Constructs a new instance of [`RefCount`].
-    pub fn new() -> Self {
-        Self {
-            count: AtomicUsize::new(1),
-        }
+impl<T: ?Sized> RefBorrow<T> {
+    /// Creates a new [`RefBorrow`] instance.
+    ///
+    /// # Safety
+    ///
+    /// Callers must ensure the following for the lifetime of the returned [`RefBorrow`] instance:
+    /// 1. That `obj` remains valid;
+    /// 2. That no mutable references to `obj` are created.
+    unsafe fn new(obj: ManuallyDrop<Ref<T>>) -> Self {
+        // INVARIANT: The safety requirements guarantee the invariants.
+        Self { inner_ref: obj }
     }
 }
 
-impl Default for RefCount {
-    fn default() -> Self {
-        Self::new()
+impl<T: ?Sized> Deref for RefBorrow<T> {
+    type Target = Ref<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.inner_ref.deref()
```

Comment:
It's a less round-about way to express the same thing: when we use `&self.inner_ref`, we're returning a `&T` where a `&U` is expected, but given that `T: Deref<Target = U>`, the compiler coerces the value into `&U` by calling `.deref()`.

When you asked the same question on another PR, I just converted to `&X` because I saw no harm in using the compiler feature to have shorter code.

However, in another discussion I saw you trying to print a pointer and using a similar technique, but in that case it didn't work because the compiler didn't have a target type to coerce to, so it just printed the address of `&T` instead of `&U`. And now I have a reason to be careful with coercion.

Reply:
You are absolutely correct, and well spotted! Let's keep the `.deref()`.

Reply:
Coercions in general make me sad too :(

And transitive coercions tend to be evil :)